modules/ingest-ecs/src/main/java/org/elasticsearch/ingest/ecs/EcsNamespaceProcessor.java

```diff
@@ -0,0 +1,275 @@
+        entry("log.level", "severity_text"),
+        entry("trace.id", "trace_id")
+    );
+
+    /**
+     * A closed-set of keys that should be kept at the top level of the processed document after applying the namespacing.
+     * In essence, these are the fields that should not be moved to the "attributes" or "resource.attributes" namespaces.
+     * Besides the @timestamp field, this set obviously contains the attributes and the resource fields, as well as the
+     * OpenTelemetry-compatible fields that are renamed by the processor.
+     */
+    private static final Set<String> KEEP_KEYS;
+    static {
+        Set<String> keepKeys = new HashSet<>(Set.of("@timestamp", "attributes", "resource"));
+        Set<String> renamedTopLevelFields = new HashSet<>();
+        for (String value : RENAME_KEYS.values()) {
+            // if the renamed field is nested, we only need to know the top level field
+            int dotIndex = value.indexOf('.');
+            if (dotIndex != -1) {
+                renamedTopLevelFields.add(value.substring(0, dotIndex));
+            } else {
+                renamedTopLevelFields.add(value);
+            }
+        }
+        keepKeys.addAll(renamedTopLevelFields);
+        KEEP_KEYS = Set.copyOf(keepKeys);
+    }
+
+    private static final String AGENT_PREFIX = "agent";
+    private static final String CLOUD_PREFIX = "cloud";
+    private static final String HOST_PREFIX = "host";
+
+    private static final String ATTRIBUTES_KEY = "attributes";
+    private static final String RESOURCE_KEY = "resource";
+    private static final String SCOPE_KEY = "scope";
+    private static final String BODY_KEY = "body";
+    private static final String TEXT_KEY = "text";
+    private static final String STRUCTURED_KEY = "structured";
+
+    EcsNamespaceProcessor(String tag, String description) {
+        super(tag, description);
+    }
+
+    @Override
+    public String getType() {
+        return TYPE;
+    }
+
+    @Override
+    public IngestDocument execute(IngestDocument document) {
+        Map<String, Object> source = document.getSource();
+
+        boolean isOTel = isOTelDocument(source);
+        if (isOTel) {
+            return document;
+        }
+
+        // non-OTel document
+
+        Map<String, Object> newAttributes = new HashMap<>();
+        // The keep keys indicate the fields that should be kept at the top level later on when applying the namespacing.
+        // However, at this point we need to move their original values (if they exist) to the one of the new attributes namespaces, except
+        // for the @timestamp field. The assumption is that at this point the document is not OTel compliant, so even if a valid top
+        // level field is found, we assume that it does not bear the OTel semantics.
+        for (String keepKey : KEEP_KEYS) {
+            if (keepKey.equals("@timestamp")) {
+                continue;
+            }
+            Object value = source.remove(keepKey);
```

Comment:
Does it matter that this will remove things that are explicitly mapped to `null` and not re-map them?

Reply:
I answered that [above](https://github.com/elastic/elasticsearch/pull/125699#discussion_r2020052854), but I think that even this doubt doesn't worth the very small optimization. I will replace with `containsKey` and maintain keys with `null` values.